### PR TITLE
Vec not required for explain prediction

### PR DESCRIPTION
**Pull request by lopuhin**
Not that I hate vectorizers :)
But ``vec`` is not strictly required for ``explain_prediction`` (in the same way it is not required for explain_weights), so I think it makes sense to have ``vec`` after ``doc``. I'm not 100% sure about 8bcb706 though - with it we need to pass vectorized=True only if we pass vec and already vectorized doc, but don't have to do it if we don't pass vec - not sure if it's too magical.

### CODE REVIEW

File: eli5/sklearn/utils.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def get_feature_names(clf, vec=None, bias_name='<BIAS>', feature_names=None):
     """
     if feature_names is None:
         if vec and hasattr(vec, 'get_feature_names'):
-            feature_names = vec.get_feature_names()
+            feature_names = list(vec.get_feature_names())
```
Comment (lopuhin) [inline eli5/sklearn/utils.py]: Forgot to add - this change is required because for some some vectorizers (DictVectorizer) get_features_names returns a list that holds the feature names for the vectorizer, and we append to it bias several lines below and modify it.

File: eli5/sklearn/explain_prediction.py [hunk 2]
```diff
@@ -111,3 +106,13 @@ def _add_intercept(X):
         return sp.hstack([X, intercept]).tocsr()
     else:
         return np.hstack([X, intercept])
+
+
+def _get_X(doc, vec=None, vectorized=False):
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: heh, I extracted exactly the same utility (not commited) and named it almost the same (`_get_X_dense`)

Comment (lopuhin) [inline eli5/sklearn/explain_prediction.py]: :) I hope I didn't change too many unrelated things! I don't have any more global changes in the pipeline, just adding prediction explanation for regressors.

File: eli5/sklearn/explain_prediction.py [hunk 3]
```diff
@@ -23,8 +23,8 @@
 
 
 @singledispatch
-def explain_prediction(clf, vec, doc, top=_TOP, class_names=None,
-                       feature_names=None, vectorized=False):
+def explain_prediction(clf, doc, vec=None, top=_TOP, class_names=None,
+                       feature_names=None, target_names=None):
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: probably it makes sense to rename class_names to target_names and keep a single argument for them; target_names is the name used in sklearn.

**Comment by kmike** [comment]
The change makes sense; I experimented with something similar, but stopped because wanted clf, vec arguments to be together, and doc can't be a first argument because of singledispatch. But it is a stupid reason to require vec argument.

File: eli5/sklearn/explain_prediction.py (hunk 3 above)
Comment (lopuhin) [inline eli5/sklearn/explain_prediction.py]: Yeah, that would be better, I like it!

**Comment by lopuhin** [comment]
Yep, clf and vec look nicer together. It shouldn't be too hard to have doc the first argument, by just introducing an intermediate function and doing singledispatch through it - what do you think?

**Comment by kmike** [comment]
> Yep, clf and vec look nicer together. It shouldn't be too hard to have doc the first argument, by just introducing an intermediate function and doing singledispatch through it - what do you think?

I think it doesn't worth it; the API in this PR looks fine.

**Comment by lopuhin** [comment]
Ok, I'll just change the class_names to target_names then.

File: eli5/sklearn/explain_prediction.py [hunk 4]
```diff
@@ -23,8 +23,8 @@
 
 
 @singledispatch
-def explain_prediction(clf, vec, doc, top=_TOP, class_names=None,
-                       feature_names=None, vectorized=False):
+def explain_prediction(clf, doc, vec=None, top=_TOP, target_names=None,
+                       feature_names=None):
```
Comment (kmike) [inline eli5/sklearn/explain_prediction.py]: why is "vectorized" argument removed?

Comment (lopuhin) [inline eli5/sklearn/explain_prediction.py]: oh, sorry - I removed it by accident, fixed now. Thanks for spotting that!

**Comment by lopuhin** [comment]
Thanks! :)
